quil-rs/src/program/analysis/control_flow_graph.rs

```diff
@@ -0,0 +1,153 @@
+                | Instruction::SetPhase(_)
+                | Instruction::SetScale(_)
+                | Instruction::ShiftFrequency(_)
+                | Instruction::ShiftPhase(_)
+                | Instruction::Store(_)
+                | Instruction::SwapPhases(_)
+                | Instruction::UnaryLogic(_)
+                | Instruction::Wait => current_block_instructions.push(instruction),
+
+                Instruction::CalibrationDefinition(_)
+                | Instruction::CircuitDefinition(_)
+                | Instruction::Declaration(_)
+                | Instruction::FrameDefinition(_)
+                | Instruction::GateDefinition(_)
+                | Instruction::Include(_)
+                | Instruction::MeasureCalibrationDefinition(_)
+                | Instruction::WaveformDefinition(_) => {}
+
+                Instruction::Jump(_)
+                | Instruction::JumpUnless(_)
+                | Instruction::JumpWhen(_)
+                | Instruction::Label(_)
+                | Instruction::Halt => {
+                    let (terminator, new_label) = match instruction {
+                        Instruction::Jump(jump) => (BasicBlockTerminator::Jump(jump), None),
+                        Instruction::JumpUnless(jump_unless) => {
+                            (BasicBlockTerminator::JumpUnless(jump_unless), None)
+                        }
+                        Instruction::JumpWhen(jump_when) => {
+                            (BasicBlockTerminator::JumpWhen(jump_when), None)
+                        }
+                        Instruction::Label(Label { target }) => {
+                            (BasicBlockTerminator::Continue, Some(target))
+                        }
+                        Instruction::Halt => (BasicBlockTerminator::Halt, None),
+                        _ => unreachable!(),
+                    };
+                    if !current_block_instructions.is_empty() {
+                        let block = BasicBlock {
+                            label: current_label.take(),
+                            instructions: std::mem::take(&mut current_block_instructions),
+                            terminator,
+                        };
+                        graph.blocks.push(block);
+                        current_block_instructions = Vec::new();
+                    }
+                    current_label = new_label;
+                }
+            }
+        }
+
+        if !current_block_instructions.is_empty() || current_label.is_some() {
+            let block = BasicBlock {
+                label: current_label.take(),
+                instructions: current_block_instructions,
+                terminator: BasicBlockTerminator::Continue,
+            };
+            graph.blocks.push(block);
+        }
+
+        graph
+    }
+
+    pub fn get_first_basic_block(&self) -> Option<BasicBlock> {
+        self.as_control_flow_graph().blocks.into_iter().next()
+    }
```

Comment:
This seems like sort of a surprising thing to do: construct the entire graph, then discard most of it. Do you think this functionality will be needed outside of tests?

Reply:
yeah this is not something we want to ship, but I do think it's quite useful as a convenience function to let the user treat the program like a block and get the metrics they want from it. Your new `TryFrom` implementation is just as good though